// ng2-screenshot/esm2015/lib/screenshot-toolbox.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, ViewChild, Inject } from '@angular/core';
import { ScreenshotComponent } from './screenshot.component';
/**
 * Custom template of screenshot toolbox, connent with screenshot
 */
let ScreenshotToolboxComponent = class ScreenshotToolboxComponent {
    constructor(parent) {
        this.parent = parent;
        this.hasContent = false;
        // const template = `<div class="screenshot-toolbox">${el.nativeElement.children().html()}</div>`;
        // renderer.selectRootElement(template);
        // el.nativeElement.remove();
    }
    cancel() {
        this.parent.cancel();
    }
    download() {
        this.parent.download();
    }
    ngAfterViewInit() {
        window.setTimeout(() => {
            this.hasContent = this.wrapper.nativeElement.children.length > 0;
        });
    }
};
ScreenshotToolboxComponent.ctorParameters = () => [
    { type: ScreenshotComponent, decorators: [{ type: Inject, args: [ScreenshotComponent,] }] }
];
tslib_1.__decorate([
    ViewChild('wrapper', { static: false }),
    tslib_1.__metadata("design:type", ElementRef)
], ScreenshotToolboxComponent.prototype, "wrapper", void 0);
ScreenshotToolboxComponent = tslib_1.__decorate([
    Component({
        selector: 'ng2-screenshot-toolbox',
        template: `
        <div class="screenshot-toolbox" [hidden]="!parent.showToolbox">
            <div #wrapper>
                <ng-content></ng-content>
            </div>
            <div *ngIf="!hasContent">
                <button class="btn" type="button" (click)="cancel()">
                    Cancel
                </button>
                <button class="btn" type="button" (click)="download()">
                    Download
                </button>
            </div>
        </div>
    `,
        styles: ["/*! Angular Screenshot - v0.1.0 - http://weihanchen.github.io/ng2-screenshot - (c) 2017 weihanchen - MIT */.screenshot-toolbox{width:100%}.screenshot-toolbox button{margin:.2em}.screenshot-toolbox .btn{position:relative;padding:.2em .4em;overflow:hidden;border-width:0;outline:0;border-radius:.2em;box-shadow:0 .1em .2em #4f4f4f;background-color:#e6e3e3;color:#4f4f4f;-webkit-transition:background-color .3s;transition:background-color .3s}.screenshot-toolbox .btn:focus,.screenshot-toolbox .btn:hover{background:#cdcccc}.screenshot-toolbox .btn>*{position:relative}.screenshot-toolbox .btn:before{content:\"\";position:absolute;top:50%;left:50%;display:block;width:0;padding-top:0;border-radius:100%;background-color:rgba(205,204,204,.3);-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%)}.screenshot-toolbox .btn:active:before{width:120%;padding-top:120%;-webkit-transition:width .2s ease-out,padding-top .2s ease-out;transition:width .2s ease-out,padding-top .2s ease-out}.screenshot-toolbox .btn-navy{background-color:#1d9e74}.screenshot-toolbox .btn-navy:focus,.screenshot-toolbox .btn-navy:hover{background-color:#138661}.screenshot-toolbox-custom{visibility:hidden}"]
    }),
    tslib_1.__param(0, Inject(ScreenshotComponent)),
    tslib_1.__metadata("design:paramtypes", [ScreenshotComponent])
], ScreenshotToolboxComponent);
export { ScreenshotToolboxComponent };
